RFC: GraphQL `@defer`

calvincestari opened the pull request:
Read the formatted RFC [here](https://github.com/apollographql/apollo-ios/blob/spike/defer/Design/3093-graphql-defer.md).

```[tasklist]
### Codegen
- [ ] https://github.com/apollographql/apollo-ios/issues/3113
- [ ] https://github.com/apollographql/apollo-ios/issues/3114
- [ ] https://github.com/apollographql/apollo-ios/issues/3139
- [ ] https://github.com/apollographql/apollo-ios/issues/3140
- [ ] https://github.com/apollographql/apollo-ios/issues/3141
- [ ] https://github.com/apollographql/apollo-ios/issues/3142
```

```[tasklist]
### Execution
- [ ] https://github.com/apollographql/apollo-ios/issues/3144
- [ ] https://github.com/apollographql/apollo-ios/issues/3145
```

```[tasklist]
### Networking
- [ ] https://github.com/apollographql/apollo-ios/issues/3118
- [ ] https://github.com/apollographql/apollo-ios/issues/3143
- [ ] https://github.com/apollographql/apollo-ios/issues/3146
- [ ] https://github.com/apollographql/apollo-ios/issues/3147
- [ ] https://github.com/apollographql/apollo-ios/issues/3148
```

```[tasklist]
### Caching
- [ ] https://github.com/apollographql/apollo-ios/issues/3149
```

calvincestari commented:
This is ready for review and comment. The section on the generated models is the weakest of the RFC right now, I will put more time into that when I get back. If you have comments/ideas for that though please comment on it and I will address it.

BoD commented on Design/3093-graphql-defer.md:
````diff
@@ -0,0 +1,326 @@
+    .fragment(EntityFragment.self, deferred: true),
+  ] }
+```
+
````
Just for a bit of context, here's how it works on Apollo Kotlin:

- fragments spread/inline fragments generate a dedicate field in the generated model
- before `@defer`, this field could be nullable if:
  - its type condition means the fields won't always be returned
  - or it has a `@skip` or `@include` (unless it's trivial)
- after `@defer`: now the field is always nullable. If it's `null` it means you haven't received this fragment yet.

Iron-Ham commented on Design/3093-graphql-defer.md:
```diff
@@ -0,0 +1,326 @@
+
+The most simple solution is to change the deferred property type to an optional version of that type. This hides detail though because you wouldn't be able to tell whether the value is `nil` because the response data has been received yet (i.e.: deferred) or whether the data was returned and it was explicitly `null`. It also gets more complicated when a type is already optional; would that result ...
+
+I explored Swift's property wrappers but they suffer from the limitation of not being able to be applied to a computed property. All model properties are computed properties because they simply route access the value in the underlying dictionary data storage. It would be nice to be able to simply annotate fragments and fields with some like `@Deferred` but it doesn't look like that is possible.
```
Not pressing: Investigate macros for this purpose?

calvincestari replied:
Macros might be a way to obscure the implementation in the generated code so the generated models wouldn't need to be regenerated if the underlying implementation changed but I don't think macros can provide a mechanism to expose the 'state' and values of deferred fragments?

AnthonyMDev replied:
Yeah, I thought about this a lot. The problem is there really isn't any amount of generated code that could really make this work. We actually want to be able to have a "metadata" accessor on a property. The only way to do this is through the `projectedValue` of a property wrapper.

We could wrap the value in some struct that had the metadata and the value, but the 99% of the time you just wanted to value and not the "state" metadata, you'd need to call `myProperty.value` which is definitely not what we want here.

We just wanted to be able to use that sweet swift syntactic sugar that only exists with property wrappers, but they have another limitation that prevents us from using them right now.

Iron-Ham commented on Design/3093-graphql-defer.md:
````diff
@@ -0,0 +1,326 @@
+  }
+```
+
+Another way which may be a bit more intuitive is to make the `server` case on `Source` have an associated value since `cache` sources will always be complete. The cache could return partial responses for deferred operations but for the initial implementation we will probably only write the cache record once all deferred fragments have been received.
````
> The cache could return partial responses for deferred operations but for the initial implementation we will probably only write the cache record once all deferred fragments have been received.

Just thinking on edge cases here. Let's imagine this scenario:
* User triggers a network request.
* One of the response fields is `@deferred` as it takes, let's say 5 seconds to resolve.
* The user navigates away from the view, ending all ongoing network requests for that view. 

In this case, there would be no cache write at all, despite the majority of data being retrieved (and valid!). This perhaps could lead to the unintended effect wherein a partial fetch that should have updated the values of various entity types across the cache don't.

AnthonyMDev replied:
Great call out @Iron-Ham. I think we're all aligned this is something we want to implement, just not for the MVP. And yes, we should clearly document this behavior.

calvincestari commented:
I've [updated](https://github.com/apollographql/apollo-ios/pull/3093/commits/323838cacc7d740283c39fac185297f62bfe56b5) the section on the generated models adding "Optional fragments" which aligns with @BoD's [comment](https://github.com/apollographql/apollo-ios/pull/3093#discussion_r1250992850) earlier.

calvincestari commented:
I'm going to be creating issues/tasks for the parts of the RFC that are less contentious and need to happen to be able to effectively support the `@defer` work.

calvincestari commented:
@AnthonyMDev / @Iron-Ham - do you have any feedback on the [completion handler](https://github.com/apollographql/apollo-ios/blob/spike/defer/Design/3093-graphql-defer.md#completion-handler) changes?

Working on the defer protocol parser today I realized there is the [`label` property](https://github.com/graphql/graphql-wg/blob/main/rfcs/DeferStream.md#payload-format) which can be part of the deferred response and I wonder if we need to feed that back to the user too or just swallow it? Is there value in passing it back to the user through the completion data?

calvincestari commented:
> Working on the defer protocol parser today I realized there is the [label property](https://github.com/graphql/graphql-wg/blob/main/rfcs/DeferStream.md#payload-format) which can be part of the deferred response and I wonder if we need to feed that back to the user too or just swallow it? Is there value in passing it back to the user through the completion data?

Thinking about this more today I don't believe there is any value in passing back the label to the user. Users work with models and models are typed which is a much stronger 'reference' than a label string. If anything it should maybe be a property on deferred fragment instances.

calvincestari commented:
@Iron-Ham, I've updated the [Generated Models](https://github.com/apollographql/apollo-ios/blob/spike/defer/Design/3093-graphql-defer.md#generated-models) section with an option (4) that would make fragments optional. I'd love to get your thoughts on it when you get a moment - thanks.

AnthonyMDev commented on Design/3093-graphql-defer.md:
```diff
@@ -0,0 +1,356 @@
+
+  public struct Data: ExampleSchema.SelectionSet {
+    public static var __selections: [ApolloAPI.Selection] { [
+      .fragment(EntityFragment?.self)
```
Below, you are showing something else. Should this be modified to:
```swift
.fragment(EntityFragment.self, deferred: true)
```

Even if we are making the fragment accessor on the model optional to indicate deferred fragments, the actual selection itself (and it's consumption in the execution layer) is going to be more clear if we specifically indicate `deferred`.

calvincestari replied:
Yes it would have the `deferred` property, these two code pieces were simply done at different times. I'll update for consistency.

AnthonyMDev commented on Design/3093-graphql-defer.md:
````diff
@@ -0,0 +1,356 @@
+}
+```
+
+4. Optional fragments (disabling field merging) - optional types are only needed when fragment fields are merged into entity selection sets. If field merging were disabled automatically for deferred fragments then the solution is simplified and we only need to alter the deferred fragments to be optional. Consuming the result data is intuitive too where a `nil` fragment value would indicate that t ...
````
I think this is probably the best approach we have. I do believe there is one specific situation where we already allow optional fragments. I think it had to do with having a conditionally included fragment with a `@include/skip` directive. We should double check that and make sure that we maintain consistency and clarity here.

calvincestari replied:
Noted - thanks.

Iron-Ham replied:
+1 -- I think this is a brilliant approach.

AnthonyMDev commented on Design/3093-graphql-defer.md:
```diff
@@ -0,0 +1,356 @@
+  case let .success(data):
+    client.fetch(query: ExampleQuery()) { result in
+      switch (result) {
+      case let .success(data):
```
I do think we should add a property to the `GraphQLResult` to indicate if the response is complete/partial as well.

AnthonyMDev commented on Design/3093-graphql-defer.md:
```diff
@@ -0,0 +1,356 @@
+      // proceed ahead on the request chain
+    }
+
+    let errorHandler: (() -> Void) = {
```
We'll just need to make sure that we have adequate testing around this to ensure these closures don't cause retain cycles. Easy to capture `self` here and cause memory leaks.

AnthonyMDev commented on Design/3093-graphql-defer.md:
```diff
@@ -0,0 +1,356 @@
+
+The initial response data and data received in each incremental response will need to be retained and combined so that each incremental response can insert the latest received incremental response data at the correct path and return an up-to-date response to the request callback.
+
+The data being retained and combined should not require another pass through the GraphQL executor though.
```
Really? I would assume it would have to go back through the executor. How do you plan on merging it without execution?

calvincestari replied:
This was mentioned purely from an efficiency point of view. If the implementation of doing that is overly complicated then yes each response will have to be executed on.

AnthonyMDev commented on Design/3093-graphql-defer.md:
```diff
@@ -0,0 +1,356 @@
+    case complete
+  }
+
+  public let response: Response
```
Oh, this is what I was referring to above! :) 

I don't like calling this `Response` though. It's not actually the response data, it's just indicating if the response is partial/complete. Think we just need some consideration on naming here.

calvincestari replied:
Noted

BobaFetters replied:
I do like the idea of providing this data to the user through an enum. As for naming, yea just `Response` maybe isn't the best, thinking about what we are trying to represent which is the status of the result, maybe something like `ResultState`? `ResponseResult` doesn't really feel right, and `ResultData` would make me think it is data and not a state representation.

calvincestari replied:
The naming we can settle on in Slack or the PR. It seems like we're aligned on the state being available at the root of `GraphQLResult` and not only on the `server` source case.

AnthonyMDev commented on Design/3093-graphql-defer.md:
````diff
@@ -0,0 +1,356 @@
+}
+```
+
+2. Another way which may be a bit more intuitive is to make the `server` case on `Source` have an associated value since `cache` sources will always be complete. The cache could return partial responses for deferred operations but for the initial implementation we will probably only write the cache record once all deferred fragments have been received. This solution becomes invalid though once th ...
````
Interesting thought!  But I agree for the reason you stated that we should stick with option 1.

AnthonyMDev commented on Design/3093-graphql-defer.md:
```diff
@@ -0,0 +1,356 @@
+
+The executor currently executes on an entire operation selection set. It will need to be adapted to be able to execute on a partial response when deferred fragments have not been received and on an isolated fragment selection set so that incremental responses can be parsed alone instead of needing to execute on the whole operation’s selection set.
+
+An alternative to parsing isolated fragment selection sets would be to execute on all the data currently received. The inefficiency with this is executing on data that would have already been executed on during prior responses.
```
Yeah, I'm skeptical that we're going to be able to get partial execution to work without a whole lot of new bugs in a reasonable amount of time. I'm sure it's possible, but I'm not sure if it's worth attempting for this 1st iteration on defer.

calvincestari commented on Design/3093-graphql-defer.md:
```diff
@@ -0,0 +1,356 @@
+
+  public struct Data: ExampleSchema.SelectionSet {
+    public static var __selections: [ApolloAPI.Selection] { [
+      .fragment(EntityFragment?.self)
```
```suggestion
      .fragment(EntityFragment?.self, deferred: true)
```

BobaFetters commented on Design/3093-graphql-defer.md:
````diff
@@ -0,0 +1,356 @@
+}
+```
+
+2. Another way which may be a bit more intuitive is to make the `server` case on `Source` have an associated value since `cache` sources will always be complete. The cache could return partial responses for deferred operations but for the initial implementation we will probably only write the cache record once all deferred fragments have been received. This solution becomes invalid though once th ...
````
Agreed, ultimately would be great to be able to get partial responses from cache but for this implementation I think option 1 is fine.

AnthonyMDev commented:
> Thinking about this more today I don't believe there is any value in passing back the label to the user. Users work with models and models are typed which is a much stronger 'reference' than a label string. If anything it should maybe be a property on deferred fragment instances.

I'm not sure about this. There may be some valuable functionality that this could unlock, but I don't think it needs to be part of the MVP either way. We can see if we get any user feedback requesting us to expose this later.

calvincestari commented:
> > Thinking about this more today I don't believe there is any value in passing back the label to the user. Users work with models and models are typed which is a much stronger 'reference' than a label string. If anything it should maybe be a property on deferred fragment instances.
> 
> I'm not sure about this. There may be some valuable functionality that this could unlock, but I don't think it needs to be part of the MVP either way. We can see if we get any user feedback requesting us to expose this later.

I initially thought there might be value until I went back and read the RFCs description/purpose of the `label` property. To me it reads that it's helpful for data patching and not so much once you have a complete response.
> `label`: `String`
> - A unique label across all `@defer` and `@stream` directives in an operation.
> - This label should be used by GraphQL clients to identify the data from patch responses and associate it with the correct fragment.
> - If provided, the GraphQL Server must add it to the payload.

calvincestari commented:
I've updated the RFC with decisions on the preferred option where there were multiple and cleaned up some of the text based on review comments.

In summary:
1. [Update graphql-js dependency](https://github.com/apollographql/apollo-ios/blob/spike/defer/Design/3093-graphql-defer.md#update-graphql-js-dependency) - option 3 is the preferred solution.
2. [Generated models](https://github.com/apollographql/apollo-ios/blob/spike/defer/Design/3093-graphql-defer.md#generated-models) - option 4 is the preferred solution.
3. [Completion handler](https://github.com/apollographql/apollo-ios/blob/spike/defer/Design/3093-graphql-defer.md#completion-handler) - option 1 is the preferred solution.
